framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java

```diff
@@ -894,7 +892,11 @@ public AnnotationFormatter getAnnotationFormatter() {
      *         empty set if no qualifiers are supported
      */
     public final Set<Class<? extends Annotation>> getSupportedTypeQualifiers() {
-        return supportedQuals;
+        if (this.supportedQuals.isEmpty()) {
+            supportedQuals.addAll(createSupportedTypeQualifiers());
```

Comment:
If you look through the source code of `Collections.unmodifieableSet(s)`, all it does is return an `UnmodifiableSet<E>` which contains an internal reference to `s`, and where it forwards getter calls while throwing exceptions for setter calls. The computation cost is an extra object that has 1 reference inside. The gain is a strong guarantee of no modification of any set that comes from `createSupportedTypeQualifiers`.

`createSupportedTypeQualifiers` is a protected method that subclasses can call directly, and making it go from immutable to mutable won't break type checkers written by the community. The decision is up to you :)

http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/7u40-b43/java/util/Collections.java#Collections.unmodifiableSet%28java.util.Set%29